python/cuda_parallel/tests/test_transform.py

```diff
@@ -0,0 +1,147 @@
+
+    got = d_out.get()
+    expected = unary_transform_host(d_in.get(), op)
+
+    np.testing.assert_allclose(expected, got, rtol=1e-5)
+
+
+def test_binary_transform(input_array):
+    def op(a, b):
+        return a + b
+
+    d_in1 = input_array
+    d_in2 = input_array
+    d_out = cp.empty_like(d_in1)
+
+    binary_transform_device(d_in1, d_in2, d_out, len(d_in1), op)
+
+    got = d_out.get()
+    expected = binary_transform_host(d_in1.get(), d_in2.get(), op)
+
+    np.testing.assert_allclose(expected, got, rtol=1e-5)
+
+
+@pytest.mark.skip(reason="https://github.com/NVIDIA/numba-cuda/issues/175")
+def test_unary_transform_struct_type():
+    import cupy as cp
+    import numpy as np
+
+    @gpu_struct
+    class MyStruct:
+        x: np.int16
+        y: np.uint64
+
+    def op(a):
+        return MyStruct(a.x * 2, a.y + 10)
+
+    num_values = 10_000
+
+    h_in = np.empty(num_values, dtype=MyStruct)
+    h_in["x"] = np.arange(num_values)
+    h_in["y"] = 1
+    d_in = cp.empty_like(h_in)
+
+    cp.cuda.runtime.memcpy(
+        d_in.data.ptr,
+        h_in.__array_interface__["data"][0],
+        h_in.nbytes,
+        cp.cuda.runtime.memcpyHostToDevice,
+    )
+
+    d_out = cp.empty_like(d_in)
+
+    transform = algorithms.unary_transform(d_in, d_out, op)
+    transform(d_in, d_out, len(d_in))
+
+    got = d_out.get()
+
+    np.testing.assert_allclose(got["x"], np.arange(num_values) * 2)
+    np.testing.assert_allclose(got["y"], np.ones(num_values) + 10)
+
+
+@pytest.mark.skip(reason="https://github.com/NVIDIA/numba-cuda/issues/175")
+def test_binary_transform_struct_type():
+    import cupy as cp
+    import numpy as np
+
+    @gpu_struct
+    class MyStruct:
+        x: np.int16
+        y: np.uint64
+
+    def op(a, b):
+        return MyStruct(a.x + b.x, a.y + b.y)
+
+    num_values = 10_000
+
+    h1_in = np.empty(num_values, dtype=MyStruct)
+    h1_in["x"] = np.random.randint(0, num_values, num_values, dtype="int16")
+    h1_in["y"] = np.random.randint(0, num_values, num_values, dtype="uint64")
+
+    h2_in = np.empty(num_values, dtype=MyStruct)
+    h2_in["x"] = np.random.randint(0, num_values, num_values, dtype="int16")
+    h2_in["y"] = np.random.randint(0, num_values, num_values, dtype="uint64")
+
+    d1_in = cp.empty_like(h1_in)
+    d2_in = cp.empty_like(h2_in)
+
+    cp.cuda.runtime.memcpy(
+        d1_in.data.ptr,
+        h1_in.__array_interface__["data"][0],
+        h1_in.nbytes,
+        cp.cuda.runtime.memcpyHostToDevice,
+    )
+    cp.cuda.runtime.memcpy(
+        d2_in.data.ptr,
+        h2_in.__array_interface__["data"][0],
+        h2_in.nbytes,
+        cp.cuda.runtime.memcpyHostToDevice,
+    )
+
+    d_out = cp.empty_like(d1_in)
+
+    transform = algorithms.binary_transform(d1_in, d2_in, d_out, op)
+    transform(d1_in, d2_in, d_out, len(d1_in))
+
+    got = d_out.get()
+
+    np.testing.assert_allclose(got["x"], h1_in["x"] + h2_in["x"])
+    np.testing.assert_allclose(got["y"], h1_in["y"] + h2_in["y"])
```

Comment:
Would be nice to have iterator and stream tests as well

Reply:
Ah good catch. Adding an iterator test revealed something we haven't seen before. Because we have two input iterators, and they can be of the same type, NVRTC compilation fails with:

```
  kernel_arg<It> arg;
  // since we switch the active member of the union, we must use placement new or construct_at. This also uses the copy
  // constructor of It, which works in more cases than assignment (e.g. thrust::transform_iterator with
  // non-copy-assignable functor, e.g. in merge sort tests)
  ::cuda::std::__construct_at(&arg.iterator, it);
  return arg;
```

(the `advance` method of both iterators are the same and thus we try to compile and link the same function twice.)

Reply:
Changing this PR to draft to address this and will re-open to review after.

Reply:
I think I ran into the same issue a while ago with merge sort. I fixed it by adding an optional prefix argument, see:

https://github.com/NVIDIA/cccl/blob/d3f1b8a828ba75b7244a89818f87a7c304e2c836/python/cuda_parallel/tests/test_merge_sort.py#L267-L269

Reply:
Sorry for missing this comment. I hacked together a solution that should enable reusing the same ltoir across iterators that have the same advance/deref operations. I believe this also avoids unnecessary recompilations

Reply:
Nader and I discussed this offline, and decided that the approach introduced in this PR should replace the `prefix=` argument.

- [this commit ](https://github.com/NVIDIA/cccl/pull/4320/commits/83d00a1fee7755a60505806683b4a1c907ec9ea7) adds a `scrub_duplicate_ltoirs` which ensures only one definition of any given advance/dereference symbol is seen by NVRTC. All other definitions are scrubbed (set to empty byte string)
- [this commit](https://github.com/NVIDIA/cccl/pull/4320/commits/5bcee9e6a6121d635a04313c0825ca2fee39f5c0) replaces the previous approach using a `prefix=` argument with ltoir scrubbing